File: app/src/main/java/im/fdx/v2ex/ui/main/MyViewPagerAdapter.java
package im.fdx.v2ex.ui.main;

import android.app.Fragment;
import android.app.FragmentManager;
import android.content.Context;
import android.os.Bundle;
import android.support.v13.app.FragmentStatePagerAdapter;

import java.util.ArrayList;
import java.util.List;

import im.fdx.v2ex.MyApp;
import im.fdx.v2ex.R;
import im.fdx.v2ex.utils.Keys;


/**
 * Created by fdx on 2015/10/15.
 * 从MainActivity分离出来. 用了FragmentStatePagerAdapter 替代FragmentPagerAdapter，才可以动态切换Fragment
 * 弃用了Volley 和 模拟web + okhttp
 */
class MyViewPagerAdapter extends FragmentStatePagerAdapter {

    private List<TopicsFragment> mFragments = new ArrayList<>();
    private List<String> mTabTitles = new ArrayList<>();
    private Context mContext;

    public MyViewPagerAdapter(FragmentManager fm, Context context) {
        super(fm);
        this.mContext = context;
        initFragment();
    }

    public synchronized void initFragment() {
            mTabTitles.clear();
            mFragments.clear();
            String[] tabTitles = mContext.getResources().getStringArray(R.array.v2ex_favorite_tab_titles);
            String[] tabPaths = mContext.getResources().getStringArray(R.array.v2ex_favorite_tab_paths);
            for (int i = 0; i < tabPaths.length; ++i) {
                TopicsFragment fragment = new TopicsFragment();
                Bundle bundle = new Bundle();
                bundle.putString(Keys.KEY_TAB, tabPaths[i]);
                fragment.setArguments(bundle);
                mFragments.add(fragment);
                mTabTitles.add(tabTitles[i]);
            }
    }

    @Override
    public Fragment getItem(int position) {
        return mFragments.get(position);
    }


    @Override
    public int getCount() {
        return mTabTitles.size();
    }

    @Override
    public CharSequence getPageTitle(int position) {
        return mTabTitles.get(position);
    }

}

File: app/src/main/java/im/fdx/v2ex/ui/main/MyViewPagerAdapter.kt
package im.fdx.v2ex.ui.main

import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.adapter.FragmentStateAdapter
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import im.fdx.v2ex.myApp
import im.fdx.v2ex.pref
import im.fdx.v2ex.ui.MyTab
import im.fdx.v2ex.ui.tabPaths
import im.fdx.v2ex.ui.tabTitles
import im.fdx.v2ex.utils.Keys
import im.fdx.v2ex.utils.Keys.PREF_TAB
import java.util.*


/**
 * Created by fdx on 2015/10/15.
 * 从MainActivity分离出来. 用了FragmentStatePagerAdapter 替代FragmentPagerAdapter，才可以动态切换Fragment
 * 弃用了Volley 和 模拟web + okhttp
 *
 * todo pageadapter有更新，明天需要完成
 */
internal class MyViewPagerAdapter(
        fa: FragmentActivity) : FragmentStateAdapter(fa) {

    private val mFragments = ArrayList<TopicsFragment>()
    val myTabList = mutableListOf<MyTab>()
    init {
        initFragment()
    }

    fun initFragment() {
        myTabList.clear()
        mFragments.clear()

        var jsonData = pref.getString(PREF_TAB, null)
        if (jsonData == null) {
            val list = MutableList(tabTitles.size) { index: Int ->
                MyTab(tabTitles[index], tabPaths[index])
            }

            jsonData = Gson().toJson(list)
        }

        val turnsType = object : TypeToken<List<MyTab>>() {}.type
        val list = Gson().fromJson<List<MyTab>>(jsonData, turnsType)

        if(list.isNullOrEmpty()) { //可能的一些奇怪的问题
            MutableList(tabTitles.size) { index: Int ->
                MyTab(tabTitles[index], tabPaths[index])
            }
        }

        for (it in list) {
            if (!myApp.isLogin && it.path == "recent") {
                continue
            }
            mFragments.add(TopicsFragment().apply { arguments = bundleOf(Keys.KEY_TAB to it.path, Keys.KEY_TYPE to it.type) })
            myTabList.add(it)
        }
    }

    override fun getItemCount(): Int {
        return  myTabList.size + 1
    }

    override fun createFragment(position: Int): Fragment {
        return if (position < itemCount -1)
            mFragments[position]
        else
            Fragment()
    }

}
